File: sdks/python/ans/integrations/openai_agents.py
"""OpenAI Agents SDK integration for ANS."""

import functools
import sys
import time
from typing import Callable

try:
    import openai  # noqa: F401
    _openai_agents_available = True
except ImportError:
    _openai_agents_available = False

from ..client import ANSClient, ANSError, hash_payload


def ans_tool_plugin(
    client: ANSClient,
    agent_id: str,
    parent_agent_id: str = "",
    action_type: str = "custom",
    silent: bool = True,
):
    """
    Decorator factory for OpenAI agent tools that adds ANS tracing.

    Args:
        client: ANS client
        agent_id: Agent ID
        parent_agent_id: Parent agent ID (if sub-agent)
        action_type: Action type for receipts
        silent: If True (default), ANSError is logged but never raised.

    Example:
        from ans.integrations.openai_agents import ans_tool_plugin
        from openai import OpenAI

        ans_client = ANSClient()
        plugin = ans_tool_plugin(ans_client, agent_id="ans_xyz")

        @plugin
        def write_file(path: str, content: str):
            with open(path, 'w') as f:
                f.write(content)
            return f"Wrote {len(content)} bytes"

        client = OpenAI()
        response = client.chat.completions.create(
            model="gpt-4",
            messages=[...],
            tools=[write_file]
        )
    """

    def _pre(ph, summary):
        try:
            resp = client.sign_append(
                agent_id=agent_id, phase="pre", action_type=action_type,
                payload_hash=ph, payload_summary=summary,
                policy_decision="allow", parent_agent_id=parent_agent_id,
            )
            return resp["receipt_id"]
        except ANSError as e:
            if not silent: raise
            print(f"ans: pre-receipt error (non-fatal): {e}", file=sys.stderr)
            return ""

    def _post(ph, summary, outcome, o_sum, dur, pre_id):
        if not pre_id: return
        try:
            client.sign_append(
                agent_id=agent_id, phase="post", action_type=action_type,
                payload_hash=ph, payload_summary=summary,
                outcome=outcome, outcome_summary=o_sum,
                duration_ms=dur, pre_receipt_id=pre_id,
                parent_agent_id=parent_agent_id,
            )
        except ANSError as e:
            if not silent: raise
            print(f"ans: post-receipt error (non-fatal): {e}", file=sys.stderr)

    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            payload = {"args": args, "kwargs": kwargs}
            ph = hash_payload(payload)
            summary = f"Tool: {func.__name__}"
            pre_id = _pre(ph, summary)

            start = time.time()
            outcome, o_sum, result, error = "success", "", None, None
            try:
                result = func(*args, **kwargs)
                o_sum = str(result)[:120] if result else "Completed"
            except Exception as e:
                outcome, o_sum, error = "failure", f"Error: {type(e).__name__}: {str(e)[:100]}", e

            _post(ph, summary, outcome, o_sum, int((time.time() - start) * 1000), pre_id)
            if error: raise error
            return result

        return wrapper

    return decorator

File: sdks/typescript/src/integrations/openai-agents.ts
/**
 * ANS integration for the OpenAI Agents JS/TS SDK.
 *
 * Provides ansAgentPlugin — a factory that wraps tool functions with ANS receipts.
 *
 * Usage:
 *   import { ansAgentPlugin } from "ans-sdk/integrations/openai-agents";
 *   import { tool } from "@openai/agents";
 *
 *   const plugin = ansAgentPlugin({ agentId: "ans_xyz123" });
 *   const wrappedFn = plugin(myTool, "my_tool_name");
 *
 * SPDX-License-Identifier: Apache-2.0
 */
import { Client, ANSError, ActionType } from "../client";

export function ansAgentPlugin(opts: {
  agentId: string;
  client?: Client;
  silent?: boolean;
}): (fn: Function, toolName: string) => Function {
  const client = opts.client || new Client();
  const silent = opts.silent !== false;

  return (fn: Function, toolName: string): Function => {
    return async (...args: unknown[]) => {
      const summary = `openai:${toolName}`.slice(0, 80);
      let preId = "";
      try {
        const resp = await client.signAppend({
          agentId: opts.agentId,
          phase: "pre",
          actionType: "custom" as ActionType,
          payload: { tool: toolName, args: JSON.stringify(args).slice(0, 200) },
          payloadSummary: summary,
          policyDecision: "allow",
          authContext: `OpenAI Agents tool: ${toolName}`,
        });
        preId = resp.receipt_id;
      } catch (e) {
        if (!silent) throw e;
        console.error(`ans: pre-receipt error: ${e instanceof ANSError ? e.message : e}`);
      }

      const start = performance.now();
      let outcome = "success";
      let resultSummary = "";
      try {
        const result = await fn(...args);
        resultSummary = result != null ? String(result).slice(0, 120) : "";
        return result;
      } catch (e) {
        outcome = "failure";
        resultSummary = e instanceof Error ? e.message.slice(0, 120) : String(e).slice(0, 120);
        throw e;
      } finally {
        const duration = Math.round(performance.now() - start);
        try {
          await client.signAppend({
            agentId: opts.agentId,
            phase: "post",
            actionType: "custom" as ActionType,
            payload: { tool: toolName },
            payloadSummary: summary,
            outcome,
            outcomeSummary: resultSummary,
            durationMs: duration,
            preReceiptId: preId,
          });
        } catch (e) {
          if (!silent) throw e;
          console.error(`ans: post-receipt error: ${e instanceof ANSError ? e.message : e}`);
        }
      }
    };
  };
}
